### ISSUE TITLE
[CPU inductor] Fails with `.view(int32)` + bit shift

### ISSUE THREAD

**Issue by gau-nernst** [pytorch/pytorch#135160]
### 🐛 Describe the bug

In torchao, we have FPx dtype conversion using bit manipulation. In recent versions of PyTorch nightly, this breaks on CPU inductor. See https://github.com/pytorch/ao/issues/792 for more background info.

See minified repro below. Comparing the codegen

```C++
// 2.4
auto tmp2 = static_cast<int32_t>(2);
auto tmp3 = decltype(tmp1)(tmp1 >> tmp2);

// nightly
auto tmp2 = static_cast<int32_t>(2);
auto tmp3 = c10::convert<float>(tmp2);
auto tmp4 = at::vec::Vectorized<float>(tmp3);
auto tmp5 = tmp1 >> tmp4;
```

I'm suspecting `c10::convert<float>` and `at::vec::Vectorized<float>` are causing issues, since we can't do bit-shift on float. Not sure why this is happening.

cc: @msaroufim @jerry

### Error logs

```
/home/user/miniconda3/envs/dev_nightly/lib/python3.10/site-packages/torch/include/ATen/cpu/vec/vec_base.h: In instantiation of ‘at::vec::CPU_CAPABILITY::Vectorized<T> at::vec::CPU_CAPABILITY::operator>>(const at::vec::CPU_CAPABILITY::Vectorized<T>&, const at::vec::CPU_CAPABILITY::Vectorized<T>&) [with T = float]’:
/tmp/torchinductor_user/hm/chmixe2p3fp5a76ikzzgavmbie7uo7mcgkcunzlqf6c3lklwqdbr.cpp:27:33:   required from here
/home/user/miniconda3/envs/dev_nightly/lib/python3.10/site-packages/torch/include/ATen/cpu/vec/vec_base.h:892:16: error: use of deleted function ‘const float& at::vec::CPU_CAPABILITY::Vectorized<float>::operator[](int) const’
  892 |     T shift = b[i];
```

### Minified repro

```python
import torch

def f(x):
    return x.view(torch.int32) >> 2

torch.compile(f)(torch.ones(16, 16))
```

If the input is int32, and we don't use `.view(dtype)` inside the compiled region, the codegen is correct (`c10::convert<int32_t>` and `at::vec::Vectorized<int32_t>`) -> no error

### Versions

```
Collecting environment information...
PyTorch version: 2.5.0.dev20240904
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-45-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4070 Ti SUPER
Nvidia driver version: 550.90.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 5 5600 6-Core Processor
CPU family:                         25
Model:                              33
Thread(s) per core:                 2
Core(s) per socket:                 6
Socket(s):                          1
Stepping:                           0
Frequency boost:                    enabled
CPU max MHz:                        4467.2852
CPU min MHz:                        2200.0000
BogoMIPS:                           7000.46
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          192 KiB (6 instances)
L1i cache:                          192 KiB (6 instances)
L2 cache:                           3 MiB (6 instances)
L3 cache:                           32 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] torch==2.5.0.dev20240904
[pip3] torchao==0.3.1
[pip3] torchaudio==2.5.0.dev20240904
[pip3] torchtune==0.0.0
[pip3] torchvision==0.20.0.dev20240904
[pip3] triton==3.0.0
[conda] blas                      1.0                         mkl  
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch-nightly
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.8           py310h5eee18b_0  
[conda] mkl_random                1.2.4           py310hdb19cb5_0  
[conda] numpy                     1.26.4          py310h5f9d8c6_0  
[conda] numpy-base                1.26.4          py310hb5e798b_0  
[conda] pytorch                   2.5.0.dev20240904 py3.10_cuda12.4_cudnn9.1.0_0    pytorch-nightly
[conda] pytorch-cuda              12.4                 hc786d27_7    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchao                   0.3.1                     dev_0    <develop>
[conda] torchaudio                2.5.0.dev20240904     py310_cu124    pytorch-nightly
[conda] torchtriton               3.0.0+dedb7bdf33           py310    pytorch-nightly
[conda] torchtune                 0.0.0                    pypi_0    pypi
[conda] torchvision               0.20.0.dev20240904     py310_cu124    pytorch-nightly
```

cc @ezyang @chauhang @penguinwu

**Comment by leslie-fang-intel** [comment]
Thanks for reporting this issue. I will take a look soon.

**Comment by leslie-fang-intel** [comment]
It's a regression introduced by https://github.com/pytorch/pytorch/pull/134569. Submit PR: https://github.com/pytorch/pytorch/pull/135301 to fix it.

**Comment by leslie-fang-intel** [comment]
Hi @gau-nernst, https://github.com/pytorch/pytorch/pull/135301 has landed please kindly help to verify.

**Comment by gau-nernst** [comment]
Thank you for the fix! I saw that you added a test identical to my snippet here. I think it should be sufficient. I will need to wait the next nightly release to check from my side right?

**Comment by leslie-fang-intel** [comment]
> I will need to wait the next nightly release to check from my side right?

Yean, I think so. Thanks.

**Comment by gau-nernst** [comment]
I can confirm the issue is fixed with `2.5.0.dev20240908`. Thank you again very much for the fix!
